newrelic/newrelic-ruby-agent
# OpenAI: instrument embeddings and chat completions

Initial PR for OpenAI instrumentation. We're merging this into the `openai_instrumentation` branch to preserve one historical PR for our initial instrumentation rollout to hopefully make it easier for others to reference in the future.

Co-authored by @hannahramadan 
Closes #2403
Closes #2404 

This PR includes:
* Basic instrumentation setup for the [`ruby-openai` gem](https://rubygems.org/gems/ruby-openai)
  * Support for versions 3.4.0+, based on versions currently in use by our customers [(dashboard)(internal)](https://onenr.io/0gR7KyqgWRo)
  * Create of an `ai` multiverse runner group
  * Add `ai`/`ruby_openai` testing to the appropriate CI workflows
* Instrumentation for `OpenAI::Client#embeddings`:
  * Record a segment on every method invocation
  * Record a metric on every method invocation 
  * Create LlmEmbedding events
* Instrumentation for `OpenAI::Client#chat`, known in the spec as chat completions:
  * Record a segment on every method invocation
  * Record a metric on every method invocation
  * Create `LlmChatCompletionSummary` events
  * Create `LlmChatCompletionMessage` events
* Add attributes to `LlmEmbedding` and `LlmChatCompletionSummary` events from the `Net::HTTP` response headers  
* Set `llm: true` as an attribute on all transaction events with LLM-related segments 
* Test using mocked responses from real OpenAI requests




It also makes the following updates to the `Llm`-namespaced classes/modules:
* Update the `LlmEvent` and other `Llm` classes to use the attribute key strings expected by the UI
* Add an `llm_event` attribute to `AbstractSegment`
* Organize attribute names to reflect the order of the spec

This PR does not contain the following GA-required tasks:
* Changelog entry
* Tests for the `Net::HTTP` response header assignment to the LLM events
* Tests that validate the OpenAI instrumentation correctly assigns LLM event attribute values
* Error tracing attributes
* Feedback API
* The overall `ai_monitoring.enabled` configuration option

These items will be opened in separate PRs that we'll merge into the `openai_instrumentation` feature branch before GA.
